extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java

```diff
@@ -365,21 +358,51 @@ BeanContainerListenerBuildItem build(
 
     @Record(RUNTIME_INIT)
     @BuildStep
-    void build(MongoClientRecorder recorder, BuildProducer<MongoClientBuildItem> mongoClients, MongodbConfig config) {
-        if (config.mongoClientConfigs != null && !config.mongoClientConfigs.isEmpty()) {
-            for (Map.Entry<String, MongoClientConfig> namedDataSourceEntry : config.mongoClientConfigs.entrySet()) {
-                String name = namedDataSourceEntry.getKey();
-                mongoClients
-                        .produce(new MongoClientBuildItem(recorder.getClient(name), recorder.getReactiveClient(name), name));
+    void configureRuntimePropertiesAndBuildClients(MongoClientRecorder recorder,
+            CodecProviderBuildItem codecProvider,
+            BsonDiscriminatorBuildItem bsonDiscriminator,
+            MongodbConfig mongodbConfig, ConfigurationBuildItem config,
+            BuildProducer<MongoClientBuildItem> mongoClients) {
+        recorder.configureRuntimeProperties(codecProvider.getCodecProviderClassNames(),
+                bsonDiscriminator.getBsonDiscriminatorClassNames(), mongodbConfig);
+
+        for (String connectionName : resolveAllConnectionNamesFromConfig(config)) {
+            mongoClients.produce(new MongoClientBuildItem(recorder.getClient(connectionName),
+                    recorder.getReactiveClient(connectionName), connectionName));
+        }
+    }
+
+    private Set<String> resolveAllConnectionNamesFromConfig(ConfigurationBuildItem config) {
```

Comment:
Why not using this new methods also in the `createMongoClientProducerBean`methods that do the same but based on Jandex ?

Reply:
That is an interesting idea! I went the way I did because that solution feels closer to the problem at hand. 
That said, I'll sleep on this idea and once the Camel folks respond with whether or not the whole idea works then I'll have settled :)

Reply:
Implemented